### PR TITLE
MM-28979 mattermod: move away from gorp to sqlx in spinmint_store.go

### PR DESCRIPTION
**Pull request by kashifsoofi**
#### Summary
mattermod: move away from gorp to sqlx in spinmint_store.go

#### Ticket Link
Fixes [15566](https://github.com/mattermost/mattermost-server/issues/15566)

#### Jira tiket
https://mattermost.atlassian.net/browse/MM-28977

### CODE REVIEW

**Comment by agnivade** [comment]
@streamer45 - As per your request :)

**Comment by agnivade** [comment]
Hi @kashifsoofi, thanks for the PR!

I have some improvements for you in a git diff format which improves 2 things:
1. We don't need to use the struct tags here because our column name and field name are the same.
2. We don't need to manually create a table. We have schema migrations to take care of that :)

```diff
diff --git a/model/spinmint.go b/model/spinmint.go
index 8cdec4d..7278b24 100644
--- a/model/spinmint.go
+++ b/model/spinmint.go
@@ -5,8 +5,8 @@ package model
 
 type Spinmint struct {
        InstanceID string `db:"InstanceId"`
-       RepoOwner  string `db:"RepoOwner"`
-       RepoName   string `db:"RepoName"`
-       Number     int    `db:"Number"`
-       CreatedAt  int64  `db:"CreatedAt"`
+       RepoOwner  string
+       RepoName   string
+       Number     int
+       CreatedAt  int64
 }
diff --git a/store/sql_spinmint_store_test.go b/store/sql_spinmint_store_test.go
index eb3f92c..43e9090 100644
--- a/store/sql_spinmint_store_test.go
+++ b/store/sql_spinmint_store_test.go
@@ -12,15 +12,6 @@ func TestSQLSpinmintStore(t *testing.T) {
        ss := getTestSQLStore(t)
 
        sms := NewSQLSpinmintStore(ss)
-       _, err := ss.dbx.Exec(`CREATE TABLE IF NOT EXISTS Spinmint (
-               InstanceId varchar(128) NOT NULL,
-               RepoOwner varchar(255) DEFAULT NULL,
-               RepoName varchar(255) DEFAULT NULL,
-               Number int(11) DEFAULT NULL,
-               CreatedAt bigint(20) DEFAULT NULL,
-               PRIMARY KEY (InstanceId)
-         ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;`)
-       require.NoError(t, err)
 
        sm := &model.Spinmint{
                RepoName: "repo-name",
diff --git a/store/sql_store.go b/store/sql_store.go
index 0d1ba62..ed9ee33 100644
--- a/store/sql_store.go
+++ b/store/sql_store.go
@@ -64,6 +64,7 @@ func initConnection(driverName, dataSource string) *SQLStore {
                        },
                },
        }
+       sqlStore.dbx.MapperFunc(func(s string) string { return s })
 
        return sqlStore
 }
diff --git a/store/sql_store_test.go b/store/sql_store_test.go
index 728840b..392cdfa 100644
--- a/store/sql_store_test.go
+++ b/store/sql_store_test.go
@@ -12,7 +12,7 @@ import (
 )
 
 const (
-       defaultMysqlDSN         = "mattermod:mattermod@tcp(localhost:3306)/mattermost_mattermod_test?charset=utf8mb4,utf8\u0026readTimeout=30s\u0026writeTimeout=30s\u0026parseTime=true"
+       defaultMysqlDSN         = "mattermod:mattermod@tcp(localhost:3306)/mattermost_mattermod_test?charset=utf8mb4,utf8&readTimeout=30s&writeTimeout=30s&parseTime=true&multiStatements=true"
        defaultMysqlRootUser    = "root"
        defaultMysqlRootUserPWD = "mattermod"
        defaultMysqlUser        = "mattermod"
@@ -52,6 +52,9 @@ func getTestSQLStore(t *testing.T) *SQLStore {
                        },
                },
        }
+       store.dbx.MapperFunc(func(s string) string { return s })
+       runMigrations(db)
+
        t.Cleanup(func() {
                if err := db.Close(); err != nil {
                        t.Fatal(err)
```

File: store/sql_spinmint_store.go [hunk 1]
```diff
@@ -15,19 +15,17 @@ type SQLSpinmintStore struct {
...
-	return s
+	return &SQLSpinmintStore{sqlStore}
 }
 
 func (s SQLSpinmintStore) Save(spinmint *model.Spinmint) (*model.Spinmint, error) {
-	if err := s.GetMaster().Insert(spinmint); err != nil {
-		if _, err := s.GetMaster().Update(spinmint); err != nil {
+	insertQuery := `INSERT INTO Spinmint (InstanceId, RepoOwner, RepoName, Number, CreatedAt)
+        VALUES (:InstanceId, :RepoOwner, :RepoName, :Number, :CreatedAt)`
+	if _, err := s.dbx.NamedExec(insertQuery, spinmint); err != nil {
```
Comment (agnivade) [inline store/sql_spinmint_store.go:24]: We can directly use the string inside the `NamedExec` call itself. Similar to how other queries are done.

**Comment by kashifsoofi** [comment]
> 
> 
> Hi @kashifsoofi, thanks for the PR!
> 
> I have some improvements for you in a git diff format which improves 2 things:
> 
>     1. We don't need to use the struct tags here because our column name and field name are the same.
> 
>     2. We don't need to manually create a table. We have schema migrations to take care of that :)
> 
> 
> ```diff
> diff --git a/model/spinmint.go b/model/spinmint.go
> index 8cdec4d..7278b24 100644
> --- a/model/spinmint.go
> +++ b/model/spinmint.go
> @@ -5,8 +5,8 @@ package model
>  
>  type Spinmint struct {
>         InstanceID string `db:"InstanceId"`
> -       RepoOwner  string `db:"RepoOwner"`
> -       RepoName   string `db:"RepoName"`
> -       Number     int    `db:"Number"`
> -       CreatedAt  int64  `db:"CreatedAt"`
> +       RepoOwner  string
> +       RepoName   string
> +       Number     int
> +       CreatedAt  int64
>  }
> diff --git a/store/sql_spinmint_store_test.go b/store/sql_spinmint_store_test.go
> index eb3f92c..43e9090 100644
> --- a/store/sql_spinmint_store_test.go
> +++ b/store/sql_spinmint_store_test.go
> @@ -12,15 +12,6 @@ func TestSQLSpinmintStore(t *testing.T) {
>         ss := getTestSQLStore(t)
>  
>         sms := NewSQLSpinmintStore(ss)
> -       _, err := ss.dbx.Exec(`CREATE TABLE IF NOT EXISTS Spinmint (
> -               InstanceId varchar(128) NOT NULL,
> -               RepoOwner varchar(255) DEFAULT NULL,
> -               RepoName varchar(255) DEFAULT NULL,
> -               Number int(11) DEFAULT NULL,
> -               CreatedAt bigint(20) DEFAULT NULL,
> -               PRIMARY KEY (InstanceId)
> -         ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4;`)
> -       require.NoError(t, err)
>  
>         sm := &model.Spinmint{
>                 RepoName: "repo-name",
> diff --git a/store/sql_store.go b/store/sql_store.go
> index 0d1ba62..ed9ee33 100644
> --- a/store/sql_store.go
> +++ b/store/sql_store.go
> @@ -64,6 +64,7 @@ func initConnection(driverName, dataSource string) *SQLStore {
>                         },
>                 },
>         }
> +       sqlStore.dbx.MapperFunc(func(s string) string { return s })
>  
>         return sqlStore
>  }
> diff --git a/store/sql_store_test.go b/store/sql_store_test.go
> index 728840b..392cdfa 100644
> --- a/store/sql_store_test.go
> +++ b/store/sql_store_test.go
> @@ -12,7 +12,7 @@ import (
>  )
>  
>  const (
> -       defaultMysqlDSN         = "mattermod:mattermod@tcp(localhost:3306)/mattermost_mattermod_test?charset=utf8mb4,utf8\u0026readTimeout=30s\u0026writeTimeout=30s\u0026parseTime=true"
> +       defaultMysqlDSN         = "mattermod:mattermod@tcp(localhost:3306)/mattermost_mattermod_test?charset=utf8mb4,utf8&readTimeout=30s&writeTimeout=30s&parseTime=true&multiStatements=true"
>         defaultMysqlRootUser    = "root"
>         defaultMysqlRootUserPWD = "mattermod"
>         defaultMysqlUser        = "mattermod"
> @@ -52,6 +52,9 @@ func getTestSQLStore(t *testing.T) *SQLStore {
>                         },
>                 },
>         }
> +       store.dbx.MapperFunc(func(s string) string { return s })
> +       runMigrations(db)
> +
>         t.Cleanup(func() {
>                 if err := db.Close(); err != nil {
>                         t.Fatal(err)
> ```
> 
> EDIT: It might also be a good idea to refactor the store initialization and the mapper to a separate function so that they can be called from the main code and test. Now especially with the mapper, it's important to keep them in sync.

Thanks @agnivade, i will have try your suggestions, i had to add the tags becuase sqlx lowercases all properties names to make column names.

**Review by agnivade — COMMENTED** [review]
Amazing stuff!

One last comment. Can we refactor

```go
sqlStore := &SQLStore{
}
sqlStore.dbx.MapperFunc(func(s string) string { return s })
```
to a separate function so that it can be called from the main code and test as well?

**Comment by kashifsoofi** [comment]
> 
> 
> Amazing stuff!
> 
> One last comment. Can we refactor
> 
> ```go
> sqlStore := &SQLStore{
> }
> sqlStore.dbx.MapperFunc(func(s string) string { return s })
> ```
> 
> to a separate function so that it can be called from the main code and test as well?

Yep, called into a meeting will do it, or can just call `initConnection` from sql_store_test.go instead of &SQLStore

**Comment by agnivade** [comment]
I think that should be doable, yeah.

**Review by agnivade — APPROVED** [review]
Amazing stuff, thank you!

**Review by streamer45 — APPROVED** [review]
Great job @kashifsoofi :+1:

**Pull request merged**